### PR TITLE
support of C# 7.0 foreach in ENC

### LINKED ISSUE #17681: Need tests to ensure edit-and-continue works with expression-bodied methods
Specifically
- [ ] expression-bodied property accessors
- [ ] expression-bodied event accessors
- [ ] expression-bodied constructors
- [ ] expression-bodied destructors
- [ ] expression-bodied operators

### LINKED ISSUE #17891: Enable EnC support for C#7 features
- [x] https://github.com/dotnet/roslyn/issues/17896 ENC support for out Var
- [x] https://github.com/dotnet/roslyn/issues/12435 ENC support for pattern matching
- [x] https://github.com/dotnet/roslyn/issues/17991 ENC support for switch
- [ ] https://github.com/dotnet/roslyn/issues/5156   ENC support for local functions
- [x] https://github.com/dotnet/roslyn/issues/17681 ENC for expression-bodied methods
- [x] https://github.com/dotnet/roslyn/issues/16960 ENC for out variable declarations
- [x] https://github.com/dotnet/roslyn/issues/12438 ENC improvements for C# 7
- [x] https://github.com/dotnet/roslyn/issues/12436 ENC for Tuples

See also https://github.com/dotnet/roslyn/issues/12438 https://github.com/dotnet/roslyn/issues/12379

### PR DESCRIPTION
**Pull request by ivanbasov**
**Customer scenario**

Enabling support of C#7.0 foreach statemets by ENC: tuples and deconstructions.
Also verified enabling manually. The only issue I see to consider is when we change tuple to deconstruction (or the opposite) as a foreach variable, it displays a rude edit saying "ENC0060	Adding 'foreach statement' around an active statement will prevent the debug session from continuing." Should we consider fixing this?

**Bugs this fixes:**

Works towards to fix #17891

**Workarounds, if any**

**Risk**
Low

**Performance impact**
Low

**Is this a regression from a previous update?**

**Root cause analysis:**

How did we miss it?  What tests are we adding to guard against it in the future?

**How was the bug found?**
Planned

### CODE REVIEW

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/LocalSlotMappingTests.cs [hunk 1]
```diff
@@ -3435,5 +3435,92 @@ .maxstack  2
 }
 ", methodToken: diff1.UpdatedMethods.Single());
         }
+
+
+        [Fact]
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/LocalSlotMappingTests.cs]: Nit: extra blank line above

**Comment by jcouv** [comment]
I suspect the PR description is incorrect. This is enabling deconstruction-foreach statement, not "expression-bodied methods".

**Comment by tmat** [comment]
:shipit:

**Comment by tmat** [comment]
>         public void ForEachVariable_Update()

Consider adding a test like this below that includes deconstruction.

---
Refers to: src/EditorFeatures/CSharpTest/EditAndContinue/ActiveStatementTests.cs:3293 in a526c48. [](commit_id = a526c485dc2a88bf9cdcf83fdbb3a654efe0ae32, deletion_comment = False)

**Comment by tmat** [comment]
>         public void ForEach_Reorder_Leaf1()

And also one similar to this.

---
Refers to: src/EditorFeatures/CSharpTest/EditAndContinue/ActiveStatementTests.cs:3331 in a526c48. [](commit_id = a526c485dc2a88bf9cdcf83fdbb3a654efe0ae32, deletion_comment = False)

**Comment by tmat** [comment]
>         public void ForEach_Update_Leaf2()

And this one :)

---
Refers to: src/EditorFeatures/CSharpTest/EditAndContinue/ActiveStatementTests.cs:3429 in a526c48. [](commit_id = a526c485dc2a88bf9cdcf83fdbb3a654efe0ae32, deletion_comment = False)

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/LocalSlotMappingTests.cs (hunk 1 above)
Comment (ivanbasov) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/LocalSlotMappingTests.cs]: Thank you! Fixed

**Comment by ivanbasov** [comment]
@jcouv, @tmat:  Thank you for comments!

1. I have added tests mentioned by @tmat. It required to make changes within ENC Analyzer. Could you please re-review?
2.  @jcouv 
> I suspect the PR description is incorrect. This is enabling deconstruction-foreach statement, not "expression-bodied methods".

It seems that these deconstruction tests together with previous changes for tuples should cover support of C# 7.0 foreach statements. @jcouv, do you see other scenarios to covers to support foreach?

3. Even edits of the variable name or type around original foreach statements are considered as rude edits. The same happens for 7.0 foreach support. Likely, we can improve this in all scenarios. Agree with @tmat to track this with a separate issue (#19154) with a lower priority.

**Comment by ivanbasov** [comment]
Will add one more test case shortly

**Comment by ivanbasov** [comment]
Ping: @tmat , @jcouv, could you please re-review with latest changes? Thanks!

**Comment by tmat** [comment]
@ivanbasov The description above links bug https://github.com/dotnet/roslyn/issues/17681 -- that's likely a mistake, this has nothing to do with expression bodied members.

**Comment by ivanbasov** [comment]
Thank you, @tmat and @jcouv! #17681 was mentioned by mistake. I could not fid a specific work item for C#7.0 foreach statement. So, now I updated a reference to a generic work item. Thanks!

File: src/EditorFeatures/CSharpTest/EditAndContinue/ActiveStatementTests.cs [hunk 2]
```diff
@@ -3469,6 +3506,49 @@ static void Main(string[] args)
...
+                }
+            }
+        }
+    }
+}";
+            var edits = GetTopEdits(src1, src2);
+            var active = GetActiveStatements(src1, src2);
+
+            edits.VerifyRudeDiagnostics(active,
+                Diagnostic(RudeEditKind.InsertAroundActiveStatement, "foreach (var c in e1)", "foreach statement"),
```
Comment (tmat) [inline src/EditorFeatures/CSharpTest/EditAndContinue/ActiveStatementTests.cs]: >"foreach statement" [](start = 94, length = 19)

Use CSharpFeaturesResources.foreach_statement

File: src/EditorFeatures/CSharpTest/EditAndContinue/ActiveStatementTests.cs [hunk 3]
```diff
@@ -3469,6 +3506,49 @@ static void Main(string[] args)
...
+                Diagnostic(RudeEditKind.InsertAroundActiveStatement, "(int b1, bool b2)", "deconstruction"),
```
Comment (tmat) [inline src/EditorFeatures/CSharpTest/EditAndContinue/ActiveStatementTests.cs]: >deconstruction [](start = 91, length = 14)

This should report "foreach statement" not "deconstruction". The error is that adding XXX around and active statement is not allowed. The deconsturction is not added around the active statement, the foreach is.

File: src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs [hunk 4]
```diff
@@ -3206,6 +3244,10 @@ private static CheckedStatementSyntax TryGetCheckedStatementAncestor(SyntaxNode
             ReportUnmatchedStatements<ForEachStatementSyntax>(diagnostics, match, (int)SyntaxKind.ForEachStatement, oldActiveStatement, newActiveStatement,
                 areEquivalent: AreEquivalentActiveStatements,
                 areSimilar: (n1, n2) => SyntaxFactory.AreEquivalent(n1.Identifier, n2.Identifier));
+
+            ReportUnmatchedStatements<ForEachVariableStatementSyntax>(diagnostics, match, (int)SyntaxKind.ForEachVariableStatement, oldActiveStatement, newActiveStatement,
```
Comment (tmat) [inline src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs]: >ReportUnmatchedStatements<ForEachVariableStatementSyntax> [](start = 12, length = 57)

We want to match regular ForEach with ForEachVeriable, so I think we need to use 
ReportUnmatchedStatements<CommonForEachStatementSyntax>

File: src/EditorFeatures/CSharpTest/EditAndContinue/ActiveStatementTests.cs [hunk 5]
```diff
@@ -3469,6 +3506,49 @@ static void Main(string[] args)
...
+                Diagnostic(RudeEditKind.InsertAroundActiveStatement, "(var a1, var a2)", "deconstruction"));
+        }
+
+        [Fact]
```
Comment (tmat) [inline src/EditorFeatures/CSharpTest/EditAndContinue/ActiveStatementTests.cs]: >[Fact] [](start = 8, length = 6)

I think we need more tests, specifically testing combinations of regular foreach statements and deconstruction foreach statements

File: src/EditorFeatures/CSharpTest/EditAndContinue/ActiveStatementTests.cs (hunk 2 above)
Comment (ivanbasov) [inline src/EditorFeatures/CSharpTest/EditAndContinue/ActiveStatementTests.cs]: thanks! fixed

File: src/EditorFeatures/CSharpTest/EditAndContinue/ActiveStatementTests.cs (hunk 3 above)
Comment (ivanbasov) [inline src/EditorFeatures/CSharpTest/EditAndContinue/ActiveStatementTests.cs]: Thank you! Fixed.

---
In reply to: [114642987](https://github.com/dotnet/roslyn/pull/19114#discussion_r114642987) [](ancestors = 114642987)

File: src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs (hunk 4 above)
Comment (ivanbasov) [inline src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs]: Thank you! Done. Extended some methods with replacing SyntaxKind parameter by Func<SyntaxNode, bool> (predicate) parameter. Please let me know if we should extend it other way.

---
In reply to: [114644314](https://github.com/dotnet/roslyn/pull/19114#discussion_r114644314) [](ancestors = 114644314)

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs [hunk 6]
```diff
@@ -7853,5 +7853,229 @@ .maxstack  2
...
+  IL_0037:  stloc.s    V_10
+  IL_0039:  ldloc.s    V_10
+  IL_003b:  ldloc.s    V_9
+  IL_003d:  ldlen
+  IL_003e:  conv.i4
+  IL_003f:  blt.s      IL_000e
+  IL_0041:  ret
+}
+");
+        }
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: Just to confirm: I assume that editing a deconstruction-foreach is still prohibited when that is the active statement. Is that correct?

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs [hunk 7]
```diff
@@ -7853,5 +7853,229 @@ .maxstack  2
...
+  .maxstack  2
+  .locals init ([unchanged] V_0,
+                [int] V_1,
+                int V_2, //x1
+                bool V_3, //y
+                double V_4, //z
+                [unchanged] V_5,
+                (int, (bool, double))[] V_6,
+                int V_7,
+                System.ValueTuple<bool, double> V_8)
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: Could you share some context on how to read this?
My understanding is that the important thing is that V_2, V_3 and V_4 were renamed.
I'm not clear why V_6, V_7 or V_8 were added. Seems like they are the same as V_0, V_1 and V_5.

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs (hunk 6 above)
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: Ok, that seems to be the case, based on reading ActiveStatementTests.cs (below, in this PR) :-)

---
In reply to: [114821359](https://github.com/dotnet/roslyn/pull/19114#discussion_r114821359) [](ancestors = 114821359)

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 8]
```diff
@@ -1703,9 +1703,22 @@ protected void AddRudeDeleteAroundActiveStatement(List<RudeEditDiagnostic> diagn
             Func<TSyntaxNode, TSyntaxNode, bool> areSimilar)
             where TSyntaxNode : SyntaxNode
         {
+            ReportUnmatchedStatements(diagnostics, match, new[] { syntaxKind }, oldActiveStatement, newActiveStatement, areEquivalent, areSimilar);
+        }
+
+            protected void ReportUnmatchedStatements<TSyntaxNode>(
```
Comment (jcouv) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs]: Formatting

**Comment by tmat** [comment]
>         protected void ReportUnmatchedStatements<TSyntaxNode>(

I don't think we need this overload - it's a complex declaration. Let's just update the current call sites and add ```new[] { ... }```

---
Refers to: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:1696 in de550be. [](commit_id = de550be83203f193fc0370bbb4a6734e57253a78, deletion_comment = False)

**Comment by tmat** [comment]
Other than a couple of details, looks good.

**Comment by ivanbasov** [comment]
>         protected void ReportUnmatchedStatements<TSyntaxNode>(

OK. Thanks!

---
In reply to: [299237324](https://github.com/dotnet/roslyn/pull/19114#issuecomment-299237324) [](ancestors = 299237324)

---
Refers to: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:1696 in de550be. [](commit_id = de550be83203f193fc0370bbb4a6734e57253a78, deletion_comment = False)

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs (hunk 8 above)
Comment (ivanbasov) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs]: Thank you!

---
In reply to: [114823061](https://github.com/dotnet/roslyn/pull/19114#discussion_r114823061) [](ancestors = 114823061)

File: src/EditorFeatures/CSharpTest/EditAndContinue/ActiveStatementTests.cs (hunk 5 above)
Comment (ivanbasov) [inline src/EditorFeatures/CSharpTest/EditAndContinue/ActiveStatementTests.cs]: This was fixed in the previous iteration. Thanks!

---
In reply to: [114644807](https://github.com/dotnet/roslyn/pull/19114#discussion_r114644807) [](ancestors = 114644807)

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs (hunk 6 above)
Comment (ivanbasov) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: Renaming variables, re-ordering embedded statements and deleting statements are supported. Those scenarios do not add new variables or change types.
Adding new loops around the active statement, changing type of variables are not supported. 

Please let me know if you expect more that supported.

---
In reply to: [114822977](https://github.com/dotnet/roslyn/pull/19114#discussion_r114822977) [](ancestors = 114822977,114821359)

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs (hunk 7 above)
Comment (ivanbasov) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueTests.cs]: Good question! Synced with Tomas. Variables without names are temporary variables. They are short lived. It is pretty hard to match them between edits. Therefore, we just do not support this presuming that they are sort lived.

---
In reply to: [114822392](https://github.com/dotnet/roslyn/pull/19114#discussion_r114822392) [](ancestors = 114822392)

**Comment by ivanbasov** [comment]
Thank you! Will send a new iteration with the changed signature.

---
In reply to: [299237619](https://github.com/dotnet/roslyn/pull/19114#issuecomment-299237619) [](ancestors = 299237619)

File: src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs [hunk 9]
```diff
@@ -3187,37 +3227,42 @@ private static CheckedStatementSyntax TryGetCheckedStatementAncestor(SyntaxNode
             // 
             // Unlike exception regions matching where we use LCS, we allow reordering of the statements.
 
-            ReportUnmatchedStatements<LockStatementSyntax>(diagnostics, match, (int)SyntaxKind.LockStatement, oldActiveStatement, newActiveStatement,
+            ReportUnmatchedStatements<LockStatementSyntax>(diagnostics, match, new [] { (int)SyntaxKind.LockStatement }, oldActiveStatement, newActiveStatement,
                 areEquivalent: AreEquivalentActiveStatements,
                 areSimilar: null);
 
-            ReportUnmatchedStatements<FixedStatementSyntax>(diagnostics, match, (int)SyntaxKind.FixedStatement, oldActiveStatement, newActiveStatement,
+            ReportUnmatchedStatements<FixedStatementSyntax>(diagnostics, match, new [] { (int)SyntaxKind.FixedStatement }, oldActiveStatement, newActiveStatement,
```
Comment (tmat) [inline src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs]: >  [](start = 83, length = 1)

Nit: no space before []

**Comment by tmat** [comment]
LGTM

File: src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs (hunk 9 above)
Comment (ivanbasov) [inline src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs]: thank you! fixed

**Comment by ivanbasov** [comment]
Tagging @MattGertz for ask mode approval

**Pull request merged**